net/socket.h

```diff
@@ -0,0 +1,259 @@
+                    debitBytes(i + 1);
+
+                    std::string outstring = oss.str();
+                    if (m_Print) {
+                        std::cout << "<<< " << outstring << std::endl;
+                    }
+                    return outstring;
+                }
+            }
+
+            // if newline is not present, append the text we received and try
+            // another read
+            oss << std::string(&m_Buffer[m_BufferStart], m_BufferBytes);
+            debitBytes(m_BufferBytes);
+        }
+    }
+
+    /*
+     * Send a buffer of specified length through the socket.
+     */
+    void send(const void *buffer, size_t len)
+    {
+        std::lock_guard<std::mutex> guard(m_SendMutex);
+        checkSocket();
+
+        int ret = ::send(m_Socket, (sendbuff_t) buffer, (bufflen_t) len, MSG_NOSIGNAL);
+        if (ret == -1) {
+            throw socket_error("Could not send " + errorMessage());
+        }
+    }
+
+    /*
+     * Send a string over the socket.
+     */
+    void send(const std::string &msg)
+    {
+        send(msg.c_str(), msg.length());
+
+        if (m_Print) {
+            std::cout << ">>> " << msg;
+        }
+    }
+
+    /*
+     * Set the current socket;
+     */
+    void setSocket(socket_t sock)
+    {
+        std::lock_guard<std::mutex> guard(m_ReadMutex);
+
+        m_Socket = sock;
+        checkSocket();
+    }
+
+private:
+    char m_Buffer[DefaultBufferSize];
+    size_t m_BufferStart = 0;
+    size_t m_BufferBytes = 0;
+    std::mutex m_ReadMutex, m_SendMutex;
+    bool m_Print;
+    socket_t m_Socket = INVALID_SOCKET;
+
+    /*
+     * Debit the byte store by specified amount.
+     */
+    void debitBytes(size_t nbytes)
+    {
+        m_BufferStart += nbytes;
+        if (m_BufferStart == DefaultBufferSize) {
+            m_BufferStart = 0;
+        }
+        m_BufferBytes -= nbytes;
+    }
+
+    /*
+     * Check that the current socket is valid.
+     */
+    void checkSocket()
+    {
+        if (m_Socket == INVALID_SOCKET) {
+            throw std::runtime_error("Bad socket " + errorMessage());
+        }
+    }
+
+    /*
+     * Make a single call to read/recv.
+     */
+    size_t readOnce(char *buffer, size_t start, size_t maxlen)
+    {
+#ifdef _MSC_VER
+        int len = recv(m_Socket, (readbuff_t) &buffer[start], (bufflen_t) maxlen, 0);
+#else
+        int len;
+        while ((len = ::read(m_Socket, (readbuff_t) &buffer[start], (bufflen_t) maxlen)) == 0) {
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+#endif
+        if (len == -1) {
+            throw socket_error("Could not read from socket " + errorMessage());
+        }
+
+        return (size_t) len;
+    }
+
+    /*
+     * Get the last error message.
+     */
+    static std::string errorMessage()
+    {
+        return " (" + std::to_string(errno) + ": " + std::strerror(errno) + ")";
+    }
+}; // Socket
+} // Net
+} // GeNNRobotics
```

Comment:
Thinking some more - I don't think a lot of this stuff belongs in socket - it should be moved up into ``Node``. When using the commands etc, the actual sockets themselves could then probably be hidden from the user entirely (which is good as you could easily fuck up the command stuff e.g. by reading too much) - all that they would then be presented with are your read and send functions etc.  I think that would also give you more scope for optimising the command system to reduce the number of runtime memory allocations if it proved necessary.